pkg/apis/core/v1/helper/helpers.go

```diff
@@ -246,6 +248,44 @@ func NodeSelectorRequirementsAsSelector(nsm []v1.NodeSelectorRequirement) (label
 	return selector, nil
 }
 
+// MatchNodeSelectorTermsORed checks whether the node labels and fields match node selector terms in ORed;
+// nil or empty term matches no objects.
+func MatchNodeSelectorTermsORed(
+	nodeSelectorTerms []v1.NodeSelectorTerm,
+	nodeLabels labels.Set,
+	nodeFields labels.Set,
+) bool {
+	for _, req := range nodeSelectorTerms {
+		labelSelector, err := NodeSelectorRequirementsAsSelector(req.MatchExpressions)
+		if err != nil {
+			glog.V(10).Infof("Failed to parse MatchExpressions: %+v,regarding as not match.", req.MatchExpressions)
+			return false
+		}
+
+		if labelSelector.Selectable() && !labelSelector.Matches(nodeLabels) {
```

Comment:
Rather than adding to the label selector interface, it seems more straightforward to do this for each term

```
if len(req.MatchExpressions) == 0 && len(req.MatchFields) == 0 {
  // empty, no match
  continue
}

if len(req.MatchExpressions) > 0 {
  labelSelector, err := …
  if err != nil {
    …
  }
  if !labelSelector.Matches(nodeLabels) {
    continue
  }
}

if len(req.MatchFields) > 0 {
  fieldSelector, err := …
  ...
  if !fieldSelector.Matches(nodeFields) {
    continue
  }
}

return true
```

Reply:
Yes, also thinking this approach. The concern is the 'duplicated' logic with `NodeSelectorRequirementsAsSelector`: `len() ==0` return `Nothing` selector, so try to avoid check len again in this func. Introduced a new func wrappers all relative logic within a struct, that's easy to update in future :)

Reply:
being explicit here is much easier to follow. The length check is not concerning…

Reply:
this comment is still outstanding